src/child.cpp

```diff
@@ -0,0 +1,473 @@
+  qscan_mode mode = none;
+  char device[128] = "\0";
+  char test[8] = "\0";
+  int speed = -1;
+  bool WT_simul = 1;
+#ifdef _WIN32
+  send(arg->connfd, IDENTV, IDENTV_LEN, 0);
+#else
+  write(arg->connfd, IDENTV, IDENTV_LEN);
+#endif
+  for (;;) {
+    ssize_t n;
+    mode = none;
+    if (term)
+      return;
+#ifdef _WIN32
+    send(arg->connfd, PROMPT, PROMPT_LEN, 0);
+#else
+    write(arg->connfd, PROMPT, PROMPT_LEN);
+#endif
+    if ((n = readline(arg->connfd, linei, MAXLINE, FD_SOCKET)) < 0) {
+      // if ((n = fgets(line, MAXLINE, arg->connfd)) == EOF) {
+      // printf("Client disconnected\n");
+      return;
+    }
+    if (n <= 1)
+      continue;
+    linei[n] = 0;
+#ifdef _WIN32
+    send(arg->connfd, "\n", 1, 0);
+#else
+    write(arg->connfd, "\n", 1);
+#endif
+    if (debug && !daemonized)
+      std::cout << inet_ntoa(arg->cliaddr.sin_addr) << ":"
+                << ntohs(arg->cliaddr.sin_port) << ": command: " << linei;
+    if (!strcmp(linei, "help\n")) {
+#ifdef _WIN32
+      send(arg->connfd, helpstr, helpstr_sz, 0);
+#else
+      write(arg->connfd, helpstr, helpstr_sz);
+#endif
+      continue;
+    } else if (!strcmp(linei, "close\n")) {
+      //          shutdown(arg->connfd, SHUT_RDWR);
+      return;
+      //      } else if (!strcmp(linei, "qhelp\n")) {
+      //          mode = help;
+    } else if (!strcmp(linei, "list\n") || !strcmp(linei, "scanbus\n")) {
+      mode = scanbus;
+    } else if (!strcmp(linei, "dinfo\n")) {
+      mode = dinfo;
+    } else if (!strcmp(linei, "minfo\n")) {
+      mode = minfo;
+    } else if (!strcmp(linei, "run\n")) {
+      mode = scan;
+    } else if (!strcmp(linei, "get\n")) {
+      std::cout << lineo << "current parameters:\ndevice: '" << device
+                << "'\ntest  : '" << test << "'\nspeed : " << speed
+                << "\nsimul : " << ((WT_simul) ? "on" : "off") << "\n";
+#ifdef _WIN32
+      send(arg->connfd, lineo, strlen(lineo), 0);
+#else
+      write(arg->connfd, lineo, strlen(lineo));
+#endif
+      continue;
+    } else if (!strncmp(linei, "set", 3)) {
+      char *linet = linei + 4;
+      size_t len;
+      if ((strlen(linei) <= 4) || linei[3] != ' ') {
+        std::cout << lineo << "QSCAND: set: needs parameter!\n";
+#ifdef _WIN32
+        send(arg->connfd, lineo, strlen(lineo), 0);
+#else
+        write(arg->connfd, lineo, strlen(lineo));
+#endif
+        continue;
+      }
+      if (!strncmp(linet, "dev=", 4)) {
+        linet += 4;
+        len = strlen(linet);
+        if (len < sizeof(device)) {
+          strncpy(device, linet, len - 1);
+        } else {
+          std::cout << lineo << "QSCAND: too long device name!\n";
+#ifdef _WIN32
+          send(arg->connfd, lineo, strlen(lineo), 0);
+#else
+          write(arg->connfd, lineo, strlen(lineo));
+#endif
+        }
+      } else if (!strncmp(linet, "test=", 5)) {
+        linet += 5;
+        len = strlen(linet);
+        if (len < sizeof(test)) {
+          strncpy(test, linet, len - 1);
```

Comment:
Codacy found an issue: [Easily used incorrectly; doesn't always \0-terminate or check for invalid pointers [MS-banned] (CWE-120).  ](https://app.codacy.com/manual/jeremiahwilson777/qpxtool/pullRequest?prid=6191247)